packages/grafana-schema/src/common/common.gen.ts

```diff
@@ -8,6 +8,37 @@
 // Run 'make gen-cue' from repository root to regenerate.
 
 
+/**
+ * These are the common properties available to all queries in all datasources.
+ * Specific implementations will *extend* this interface, adding the required
+ * properties for the given context.
+ */
+export interface DataQuery {
+  /**
+   * For mixed data sources the selected datasource is on the query level.
+   * For non mixed scenarios this is undefined.
+   * TODO find a better way to do this ^ that's friendly to schema
+   */
+  datasource?: DataSourceRef;
```

Comment:
Does this `TODO` relate to the fact that `datasource` can no longer be `null`, just `undefined` | `DataSourceRef`?

Reply:
no, we're not actually enforcing the no-nulls rule yet, though...needs to happen soon.

It's about the mixed datasource type busting up clear abstraction layers. This field (target-level) is _only_ expected to be populated when when the mixed datasource is in use at the panel level. Thus, we're effectively leaking information from the container/dashboard level down into the plugin.